yarn-project/p2p/src/services/tx_collection/fast_tx_collection.ts

```diff
@@ -0,0 +1,329 @@
+        if (batch.length === 0) {
+          return;
+        }
+
+        // Collect this batch from the node
+        await this.collectionManager.collect(
+          txHashes => node.getTxsByHash(txHashes),
+          batch.map(({ txHash }) => TxHash.fromString(txHash)),
+          {
+            description: `fast ${node.getInfo()}`,
+            node: node.getInfo(),
+            method: 'fast-node-rpc',
+            ...request.blockInfo,
+          },
+        );
+
+        // Clear from the active requests the txs we just requested
+        for (const requestedTx of batch) {
+          activeRequestsToThisNode.delete(requestedTx.txHash);
+        }
+
+        // Sleep a bit until hitting the node again (or not, depending on config)
+        if (notFinished()) {
+          await sleep(this.config.txCollectionFastNodeIntervalMs);
+        }
+      }
+    };
+
+    // Kick off N parallel requests to the node, up to the maxParallelRequests limit
+    await Promise.all(times(maxParallelRequests, processBatch));
+  }
+
+  private async collectFastViaReqResp(request: FastCollectionRequest, opts: { pinnedPeer?: PeerId }) {
+    const timeoutMs = +request.deadline - this.dateProvider.now();
+    const pinnedPeer = opts.pinnedPeer;
+    const maxPeers = boundInclusive(Math.ceil(request.missingTxHashes.size / 2), 8, 32);
+    const maxRetryAttempts = 5;
+    const blockInfo = request.blockInfo;
+    const slotNumber = blockInfo.slotNumber;
+
+    if (timeoutMs < 100) {
+      this.log.warn(
+        `Not initiating fast reqresp for txs for ${request.type} at slot ${blockInfo.slotNumber} due to timeout`,
+        { timeoutMs, ...blockInfo },
+      );
+      return;
+    }
+
+    this.log.debug(
+      `Starting fast reqresp for ${request.missingTxHashes.size} txs for ${request.type} at slot ${blockInfo.slotNumber}`,
+      { ...blockInfo, timeoutMs, pinnedPeer },
+    );
+
+    try {
+      await this.collectionManager.collect(
+        txHashes =>
+          this.reqResp.sendBatchRequest<ReqRespSubProtocol.TX>(
+            ReqRespSubProtocol.TX,
+            txHashes,
+            pinnedPeer,
+            timeoutMs,
+            maxPeers,
+            maxRetryAttempts,
+          ),
+        Array.from(request.missingTxHashes).map(txHash => TxHash.fromString(txHash)),
+        { description: `reqresp for slot ${slotNumber}`, method: 'fast-req-resp', ...opts, ...request.blockInfo },
+      );
+    } catch (err) {
+      this.log.error(`Error sending fast reqresp request for txs`, err, {
+        txs: [...request.missingTxHashes],
+        ...blockInfo,
+      });
+    }
+  }
+
+  /**
+   * Handle txs by marking them as found for the requests that are waiting for them, and resolves the request if all its txs have been found.
+   * Called internally and from the main tx collection manager whenever the tx pool emits a tx-added event.
+   */
+  public foundTxs(txs: TxWithHash[]) {
```

Comment:
IIUC this gets called by the parent class? I find the the flow of data a bit hard to follow because it creates a loop tx collection -> fast/slow strategy -> tx collection (to track requests) -> fast/slow strategy (notify)

Reply:
I've added a `CollectionSink` to try and make it clearer